torch_xla/core/dynamo_bridge.py

```diff
@@ -351,6 +343,102 @@ def optimized_mod(*args):
 
     none_remover.add_nones(result)
-    return result
+    if len(result) == 1:
+      return result[0]
+    else:
+      return result
 
   return optimized_mod
+
+
+class FallBackNodeCollector(torch.fx.Interpreter):
+
+  def __init__(self, module):
+    super().__init__(module)
+    self._fallback_ops = []
+
+  def run_node(self, n: torch.fx.Node):
+    metrics.clear_counters()
+    result = super().run_node(n)
+    fallback_ops = get_fallback_ops()
+    if len(fallback_ops) > 0:
+      self._fallback_ops.append(n)
+    else:
+      # if inputs are non-xla tensors, it should be executed on CPU
+      if n.op in ["call_function", "call_module"]:
+        args, kwargs = self.fetch_args_kwargs_from_env(n)
+        for arg in args:
+          if isinstance(arg, torch.Tensor) and not is_xla_tensor(arg):
+            self._fallback_ops.append(n)
+            break
+    return result
+
+  def get_fallback_ops(self):
+    return self._fallback_ops
+
+
+class InputCollector(torch.fx.Interpreter):
+
+  def __init__(self, module):
+    super().__init__(module)
+
+  def call_module(self, target, args, kwargs):
+    if "fused_" in target:
+      submod = self.fetch_attr(target)
+      submod.xla_args = args
+    return super().call_module(target, args, kwargs)
+
+
+def extract_compiled_graph(xla_model, xla_args):
+  # This logic, needed for supporting in-place operations, is a duplicate of
+  # the one in the main `extract_internal` function above. We need to do this
+  # check for fetching fallback ops as well.
+  # TODO (@wonjoo): Make this duplicate code a bit cleaner.
+  xla_args_need_update_bool = torch_xla._XLAC._check_tensor_need_materialization(
+      xla_args)
+
+  cloned_xla_args = [
+      torch.clone(xla_arg) if isinstance(xla_arg, torch.Tensor) else xla_arg
+      for xla_arg in xla_args
+  ]
+
+  # execute model once to collect fallback ops
+  collector = FallBackNodeCollector(xla_model)
+  collector.run(*xla_args)
+  fallback_ops = collector.get_fallback_ops()
+
+  class XlaOperatorSupport(torch.fx.passes.operator_support.OperatorSupport):
+
+    def is_node_supported(self, submodules, node: torch.fx.Node) -> bool:
+      return node.op in [
+          "call_function", "call_module"
+      ] and (node not in fallback_ops or node.target == operator.getitem)
+
+  # partition the model and exectue to collect inputs
+  supported_ops = XlaOperatorSupport()
+  partitioner = CapabilityBasedPartitioner(xla_model, supported_ops)
+  partitions = partitioner.propose_partitions()
+  partitioned_graph = partitioner.fuse_partitions(partitions)
+  InputCollector(partitioned_graph).run(*xla_args)
+
+  # Again, same logic in the `extract_internal` above to support in-place operations.
+  # TODO (@wonjoo): Make this duplicate code a bit cleaner.
+  if xla_args_need_update_bool:
+    for xla_arg, cloned_xla_arg in zip(xla_args, cloned_xla_args):
+      if isinstance(xla_arg, torch.Tensor):
+        xla_arg.copy_(cloned_xla_arg)
+
```

Comment:
I think if you call `ClearPendingIr` here you will avoid the unncessary `CompileTime` and `ExecuteTime`.